syntax = "proto3";

package raffle;

service RaffleService {
    rpc PurchaseTickets (PurchaseTicketsRequest) returns (PurchaseTicketsResponse);
    rpc GetCurrentRaffle (GetCurrentRaffleRequest) returns (GetCurrentRaffleResponse);
    rpc IncreaseBalance (IncreaseBalanceRequest) returns (IncreaseBalanceResponse);
    rpc GetRaffleResults (GetRaffleResultsRequest) returns (GetRaffleResultsResponse);
}

message IncreaseBalanceRequest {
    string user_id = 1;
    int32 amount = 2;
}

message IncreaseBalanceResponse {
    int32 balance = 1;
}

message PurchaseTicketsRequest {
    string user_id = 1;
    int32 ticket_count = 2;
    string game_type = 3;
    optional string toNumber = 4;
}

message PurchaseTicketsResponse {
    repeated string ticket_numbers = 1;
    string raffle_id = 2;
}

message GetCurrentRaffleRequest {
    string user_auth_token = 1;
    string game_type = 2;
}

message GetCurrentRaffleResponse {
    string raffle_id = 1;
    int64 end_time = 2;
    bool participating = 3;
    double current_pool = 4;

    oneof raffle_specific_data {
        Lucky31Data lucky_31_data = 5;
        LuckyRaffleData lucky_raffle_data = 6;
    }
}

message Lucky31Data {
    map<int64, int64> ticket_numbers = 1;
}

message LuckyRaffleData {
}

message GetRaffleResultsRequest {
}

message Winner {
    string user_id = 1;
    double amount = 2;
    int32 position = 3;
}

message GetRaffleResultsResponse {
    string raffle_id = 1;
    int64 start_time = 2;
    int64 end_time = 3;
    double total_pool = 4;
    repeated Winner winners = 5;
}